service/forward_service.cc

```diff
@@ -370,34 +556,42 @@ future<query::forward_result> forward_service::dispatch(query::forward_request r
                     flogger.debug("dispatching forward_request={} to address={}", req_with_modified_pr, addr);
 
-                    query::forward_result partial_result = co_await dispatcher_.dispatch_to_node(
-                        addr,
-                        req_with_modified_pr
-                    );
-
-                    query::forward_result::printer partial_result_printer{
-                        .types = req_with_modified_pr.reduction_types,
-                        .res = partial_result
-                    };
-                    tracing::trace(tr_state_, "Received forward_result={} from {}", partial_result_printer, addr);
-                    flogger.debug("received forward_result={} from {}", partial_result_printer, addr);
-
-                    if (result_) {
-                        result_->merge(partial_result, req_with_modified_pr.reduction_types);
-                    } else {
-                        result_ = partial_result;
-                    }
+                    return dispatcher_.dispatch_to_node(addr, req_with_modified_pr).then(
+                        [&req, addr = std::move(addr), &result_, tr_state_ = std::move(tr_state_)] (
+                            query::forward_result partial_result
+                        ) mutable {
+                            forward_aggregates aggrs(req);
+                            query::forward_result::printer partial_result_printer{
+                                .functions = get_functions(req),
+                                .res = partial_result
+                            };
+                            tracing::trace(tr_state_, "Received forward_result={} from {}", partial_result_printer, addr);
+                            flogger.debug("received forward_result={} from {}", partial_result_printer, addr);
+                            
+                            return aggrs.with_thread_if_needed([&result_, &aggrs, partial_result = std::move(partial_result)] () mutable {
+                                if (result_) {
+                                    aggrs.merge(*result_, partial_result);
+                                } else {
+                                    result_ = partial_result;
+                                }
+                            });
+                    });       
                 }
             ).then(
-                [&result, &req, &tr_state] () -> query::forward_result {
-                    query::forward_result::printer result_printer{
-                        .types = req.reduction_types,
-                        .res = *result
-                    };
-                    tracing::trace(tr_state, "Merged result is {}", result_printer);
-                    flogger.debug("merged result is {}", result_printer);
-
-                    return *result;
+                [&result, &req, &tr_state] () -> future<query::forward_result> {
+                    forward_aggregates aggrs(req);
```

Comment:
It sounds a little wasteful to create the forward_aggregates class by copying all the vectors, just in order to use its `with_thread_if_needed` function, and then discard it and recreate the same object again in line 584 to avoid liveness issues... but on the other hand, these vectors are small and this operation is rather cheap compared to everything else that happens on this path, so perhaps it's good enough to leave it as is.

Reply:
Oh wait, no, it can be easily done. At the beginning of this function (line 536) you use `do_with` anyway. That's where you can also create the forward_aggregates, right? Since it only needs a request to be initialized, and you have `req` available - e.g. you could create `forward_aggregates aggrs(req)` right before line 536, and then pass the forward aggregates instance to do_with, so that lives long enough. And then you can pass them via reference to `with_thread_if_needed`.

Reply:
I've wrapped `parallel_for_each`'s continuation with `do_with`. I cannot pass forward aggregates to `do_with` you mention, since separate forward_aggregates are needed in `parallel_for_each` (line 544).

Reply:
Why not a coroutine?

Reply:
There was a coroutine, but I've made a mistake like one described here https://github.com/scylladb/scylla/pull/10602#issuecomment-1166631581, so I prefer to stay with continuations to avoid this